RFC-0026-logging-system.md

```diff
@@ -0,0 +1,103 @@
+
+  - **Beta**: Emitted when a beta feature is called. See
+    [PyTorch feature classifications](https://pytorch.org/blog/pytorch-feature-classification-changes/)
+
+  - **Prototype**: Emitted when a prototype feature is called. See
+    [PyTorch feature classifications](https://pytorch.org/blog/pytorch-feature-classification-changes/)
+
+* Continue using warning/error APIs that currently exist in PyTorch wherever
+  possible. For instance, `TORCH_CHECK`, `TORCH_WARN`, and `TORCH_WARN_ONCE`
+  should continue to be used in C++
+
+  - NOTE: These existing APIs don't currently have a concept of message classes,
+    so that will need to be added
+
+* Creating new message classes and severity levels should be easy
+
+* Settings to turn warnings for a specific message class into errors
+
+* Settings to disable specific message classes and severity levels
+
+  - TODO: However, most errors should not be disableable, right? Perhaps only
+    some message classes should allow disabling or downgrading errors. For
+    instance, currently in PyTorch, we can downgrade a nondeterministic error
+    to a warning, but we wouldn't want to downgrade an error from invalid
+    arguments given to an operation.
+
+* Settings to avoid emitting duplicate messages generated by multiple
+  `torch.distribted` ranks (related to issue
+  [#68768](https://github.com/pytorch/pytorch/issues/68768))
+
+* Ability to make a particular warning only warn once
+
+  - NOTE: Currently `TORCH_WARN_ONCE` does this in C++, but there is no Python
+    equivalent
+
+  - TODO: Should there be a setting to turn a warn-always into a warn-once for
+    a given message class and vice versa?
+
+  - TODO: Should warn-once be its own separate severity level?
+
+* Settings can be changed from Python, C++, or environment variables
+
+  - Filtering warnings with Python command line arguments should
+    remain possible. For instance, the following turns a `DeprecationWarning`
+    into an error: `python -W error::DeprecationWarning your_script.py`
+
+* Should integrate with Meta's internal logging system, which is
+  [glog](https://github.com/google/glog)
+
+  - TODO: What are all the requirements that define "integrating with glog"
+
+* Must be OSS-friendly, so it shouldn't require libraries (like glog) which may
+  cause incompatibility issues for projects that use PyTorch
+
+* TODO: Determine the requirements for the following concepts:
+
+  - Log files (default behavior and any settings)
+
+
+## **Motivation**
+Original issue: [link](https://github.com/pytorch/pytorch/issues/72948)
+
+Currently, it is challenging for PyTorch developers to provide messages that
+act consistently between Python and C++.
+
+It is also challenging for PyTorch users to manage the messages that PyTorch
+emits. For instance, if a PyTorch user happens to be calling PyTorch functions
+that emit lots of warnings, it can be difficult for them to filter out those
```

Comment:
python warning filter should already work for that?

Reply:
Using existing Python features wherever possible is a great idea (although it doesn't help C++ API users, but we could choose not to support them at this time). We should provide documentation for how to handle common scenarios either way. I'm not sure how Python warning filters work with various logging systems.

Reply:
I don't think the two are exclusive:
If the "catching" code is in python, then python-provided filters should work.
If the "catching" code is in c++, there is no existing system and we could build one (I would agree with @mruberry though that this is most likely low priority).

In any case, all warning thrown from python or c++ can be caught in python.
All warnings thrown from c++ can be caught in c++.

Reply:
Makes sense, the Python warning filter seems good enough.

However, on a related note, it would also be good to have a way to filter out **info** severity messages, which are just `print` calls at the moment. Related to https://github.com/pytorch/pytorch/issues/68768